lsp-mode.el

```diff
@@ -5703,6 +5868,141 @@ or `(point)' lies outside `lsp--semantic-highlighting-region'.")
+             (line-delta)
+             (column 0)
+             (face)
+             (ov)
+             (line-start-pos)
+             (final-beg)
+             (final-end)
+             (line-min)
+             (line-max-inclusive)
+             (modifier-bitmask)
+             (modifier-index)
+             (modifier-cur-bit)
+             (text-property-beg)
+             (text-property-end))
+        (save-mark-and-excursion
+          (save-restriction
+            (widen)
+            (goto-char beg)
+            (goto-char (line-beginning-position))
+            (setq final-beg (point))
+            (setq line-min (line-number-at-pos))
+            (with-silent-modifications
+              (goto-char end)
+              (goto-char (line-end-position))
+              (setq final-end (point))
+              (setq line-max-inclusive (line-number-at-pos))
+              (forward-line (- line-min line-max-inclusive))
+              (let ((skip-lines (- line-min current-line)))
+                (while (and (<= i0 i-max) (< (aref data i0) skip-lines))
+                  (setq skip-lines (- skip-lines (aref data i0)))
+                  (setq i0 (+ i0 5)))
+                (setq current-line (- line-min skip-lines)))
+              (forward-line (- current-line line-min))
+              (setq line-start-pos (point))
+              (cl-loop
+               for i from i0 to i-max by 5 do
+               (setq line-delta (aref data i))
+               (unless (= line-delta 0)
+                 (forward-line line-delta)
+                 (setq line-start-pos (point))
+                 (setq column 0)
+                 (setq current-line (+ current-line line-delta)))
+               (setq column (+ column (aref data (1+ i))))
+               (setq face (aref faces (aref data (+ i 3))))
+               (setq text-property-beg (+ line-start-pos column))
+               (setq text-property-end (+ text-property-beg (aref data (+ i 2))))
+               (when face (put-text-property text-property-beg text-property-end 'face face))
+               (cl-loop for i from 0 to (1- (length modifier-faces)) do
+                        (when (and (aref modifier-faces i)
+                                   (> 0 (logand (aref data (+ i 4)) (lsh 1 i))))
+                          (add-face-text-property text-property-beg text-property-end
+                                                  (aref modifier-faces i))))
+               when (> current-line line-max-inclusive) return nil)))))
+      `(jit-lock-bounds ,beg . ,end))))
+
+(defun lsp--semantic-tokens-request ()
+  (let ((cur-version lsp--cur-version))
+    (lsp-request-async
+     "textDocument/semanticTokens"
+     `(:textDocument ,(lsp--text-document-identifier))
+     (lambda (response)
+       (setq lsp--semantic-tokens-cache response)
+       (puthash "documentVersion" cur-version lsp--semantic-tokens-cache)
+       (font-lock-flush))
+     :mode 'tick
+     :cancel-token :semantic-tokens)))
```

Comment:
I think that here we want to have cancel token that includes the buffer name - otherwise, if another buffer performs this call it will cancel that one.

Reply:
good point, I replaced this by a string that contains `lsp--buffer-uri` for now. Or maybe you had something else in mind?

Reply:
this is exactly what I was looking for.